src/Components/AdvancedSearch/Internal/QueryBuilder/QueryBuilder.tsx

```diff
@@ -71,67 +123,46 @@ const QueryBuilder: React.FC<IQueryBuilderProps> = (props) => {
         validityMap.current.set(rowId, false);
         setIsSearchDisabled(!checkIsValidQuery());
-    };
-
-    const removeRow = (index: number, rowId: string) => {
-        // Remove row component
-        const newRows = [...rows];
-        if (index === 0) {
-            newRows.shift();
-        } else {
-            newRows.splice(index, 1);
-        }
-        updateRows(newRows);
-        // Remove row values in query and property
-        querySnippets.current.delete(rowId);
-        propertyNames.current.delete(rowId);
-        validityMap.current.delete(rowId);
-        setIsSearchDisabled(!checkIsValidQuery());
-    };
-
-    const onChangeValue = (rowId: string, rowValue: QueryRowType) => {
-        if (rowValue.value.length) {
-            validityMap.current.set(rowId, true);
-        } else {
-            validityMap.current.set(rowId, false);
-        }
-        querySnippets.current.set(rowId, rowValue);
-        setIsSearchDisabled(!checkIsValidQuery());
-    };
-
-    const onChangeProperty = (
-        rowId: string,
-        propertyName: string,
-        propertyType: PropertyValueType
-    ) => {
-        // Either create new property entry or modify existing one
-        propertyNames.current.set(rowId, propertyName);
-        const propertyNameArray = propertyNames.current.values();
-        // Then send to parent component
-        updateColumns(new Set<string>(propertyNameArray));
-        // Reset validity, string is the only field with no default value and blank is not a valid input
-        if (propertyType !== 'string') {
-            validityMap.current.set(rowId, true);
-        } else {
-            validityMap.current.set(rowId, false);
-        }
-        setIsSearchDisabled(!checkIsValidQuery());
-    };
+    }, [rows]);
 
-    const onSearch = () => {
+    const onSearch = useCallback(() => {
         const query = buildQuery(Array.from(querySnippets.current.values()));
         executeQuery(query);
+    }, [executeQuery]);
+
+    // Non-useCallback functions
+    const checkIsValidQuery = () => {
+        if (validityMap.current.size === 0) {
+            return true;
+        }
+        let isValid = true;
+        validityMap.current.forEach((validity: boolean) => {
+            if (!validity) {
+                isValid = false;
+            }
+        });
+        return isValid;
     };
 
     return (
         <div className={classNames.root}>
-            <Stack horizontal={true}>
-                <h4 className={classNames.columnHeader}>And/Or</h4>
-                <h4 className={classNames.firstColumnHeader}>Property</h4>
-                <h4 className={classNames.columnHeader}>Operator</h4>
-                <h4 className={classNames.columnHeader}>Value</h4>
-            </Stack>
-            <Stack style={{ maxHeight: 150, overflow: 'auto' }}>
+            <div className={classNames.headerGrid}>
+                {!(rows.length === 1) && (
+                    <p className={classNames.headerText}>
```

Comment:
do we need to put the class on every time? Can't we just style the container? I suspect you're overriding spacing with this class which wouldn't be necessary if these were just `span` instead of `p` tags.

Reply:
span tags don't interact well with grid, they ignore it.

Reply:
if they are styled by the grid, a `div` would work well then I think no? Then you don't have to override the spacing that comes with `p` tags.

Reply:
Isn't `p` a better html element semantically speaking?